waiter/src/waiter/scheduler/composite.clj

```diff
@@ -0,0 +1,128 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns waiter.scheduler.composite
+  (:require [clojure.string :as str]
+            [plumbing.core :as pc]
+            [schema.core :as s]
+            [waiter.scheduler :as scheduler]
+            [waiter.util.utils :as utils]))
+
+(defrecord CompositeScheduler [service-id->scheduler scheduler-id->scheduler]
+
+  scheduler/ServiceScheduler
+
+  (get-service->instances [_]
+    (->> (vals scheduler-id->scheduler)
+         (map #(future (scheduler/get-service->instances %)))
+         (map deref)
+         (apply merge)))
```

Comment:
I think this is too lazy. I think we should be forcing these values to ensure that the queries go off to the scheduler immediately.

Also, why use `(->> ... (map #(future ...)) (map deref) ...)` rather than `(->> ... (pmap ...) ...)`?

```clojure
(->> (vals scheduler-id->scheduler)
     (pmap scheduler/get-service->instances)
     doall
     (reduce into {})))
```

Reply:
Sounds good.